pkg/ipcache/kvstore.go

```diff
@@ -292,15 +293,41 @@ func findShadowedCIDR(pair *identity.IPIdentityPair) (identity.NumericIdentity,
 	return IPIdentityCache.LookupByIP(cidrStr)
 }
 
-func ipIdentityWatcher(listeners []IPIdentityMappingListener) {
-	log.Info("Starting IP identity watcher")
+// IPIdentityWatcher is a watcher that will notify when IP<->identity mappings
+// change in the kvstore
+type IPIdentityWatcher struct {
+	backend kvstore.BackendOperations
+	stop    chan struct{}
+}
 
-	for {
+// NewIPIdentityWatcher creates a new IPIdentityWatcher using the specified
+// kvstore backend
+func NewIPIdentityWatcher(backend kvstore.BackendOperations) *IPIdentityWatcher {
+	watcher := &IPIdentityWatcher{
+		backend: backend,
+		stop:    make(chan struct{}),
+	}
 
-		watcher := kvstore.ListAndWatch("endpointIPWatcher", IPIdentitiesPath, 512)
+	return watcher
+}
+
+// Run starts the watcher and blocks waiting for events. When events are
+// received from the kvstore, All IPIdentityMappingListener are notified. The
+// function returns when IPIdentityWatcher.Close() is called. The watcher will
+// automatically restart as required.
+func (iw *IPIdentityWatcher) Run(listeners []IPIdentityMappingListener) {
+restart:
+	watcher := iw.backend.ListAndWatch("endpointIPWatcher", IPIdentitiesPath, 512)
 
+	for {
+		select {
 		// Get events from channel as they come in.
-		for event := range watcher.Events {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				log.Debugf("%s closed, restarting watch", watcher.String())
+				time.Sleep(500 * time.Millisecond)
+				goto restart
```

Comment:
I'm pretty sure we don't need to call `watcher.Stop()` here but does it need any other cleanup?

Reply:
This is existing code. I've only moved it to be able to also select on the stop channel. TBH, looking closer, this is even dead code as the only condition that `.Events` can be closed is when `Stop()` is called. The logic to re-create the watcher when the real etcd channel is closed is implemented in the kvstore logic itself.